### ISSUE TITLE
DESeq2 error:   argument 1 is not a vector

### ISSUE THREAD

**Issue by bioinfonm** [joey711/phyloseq#444]
Hello, 

A few weeks ago I was running DESeq2 and then using the tutorial here:
https://github.com/joey711/phyloseq/issues/283
to make a heatmap of the variance stabilized data. 

I updated my version of phyloseq to 1.9.15 and now I am getting this error when I try to plot the heat map.

My DESeq2 follows the tutorial and the differential expression plotting works just fine. 

What follows is for plotting the heatmap
```
> diagdds = estimateSizeFactors(diagdds)
> diagdds = estimateDispersions(diagdds)
found already estimated dispersions, replacing these
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
> diagvst = getVarianceStabilizedData(diagdds)
> dim(diagvst)
[1] 6302   29
> otu_table(phylo0) <- otu_table(diagvst, taxa_are_rows=TRUE)
```
```
> phylo0
phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 6302 taxa and 29 samples ]
sample_data() Sample Data:       [ 29 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 6302 taxa by 8 taxonomic ranks ]
> keepOTUs = rownames(res[res$padj < alpha, ])
> phylo_trim0 = prune_taxa(keepOTUs, phylo0)
```
```
> otu_table(phylo_trim0)[1:10, 1:5]
OTU Table:          [10 taxa and 5 samples]
                     taxa are rows
                   S4       V13           C1        S10       GE2
denovo1770 -3.0803033 -3.080303 -3.080303271  0.9016734 -3.080303
denovo425  -3.0803033 -3.080303 -3.080303271 -0.3872257 -3.080303
denovo427  -3.0803033 -3.080303 -3.080303271  2.9790234 -3.080303
denovo1752 -3.0803033 -3.080303 -3.080303271  2.5105003 -3.080303
denovo3343 -3.0803033 -3.080303 -3.080303271  0.9016734 -3.080303
denovo675  -3.0803033 -3.080303 -3.080303271  0.9016734 -3.080303
denovo445   0.1182303 -3.080303 -3.080303271  3.1663252 -3.080303
denovo441   0.1182303 -3.080303 -0.002176105  1.2728267 -3.080303
denovo1731 -3.0803033 -3.080303 -3.080303271 -3.0803033 -3.080303
denovo381  -3.0803033 -3.080303 -3.080303271  2.5105003 -3.080303
```
```
> tax_table(phylo_trim0)[1:10, 1:5]
Taxonomy Table:     [10 taxa by 5 taxonomic ranks]:
           Kingdom    Phylum           Class                  Order                Family               
denovo1770 "Bacteria" "Proteobacteria" "Deltaproteobacteria"  "Myxococcales"       NA                   
denovo425  "Bacteria" "Bacteroidetes"  "Sphingobacteriia"     "Sphingobacteriales" "Sphingobacteriaceae"
denovo427  "Bacteria" "Actinobacteria" "Actinobacteria"       "Actinomycetales"    "Geodermatophilaceae"
denovo1752 "Bacteria" "Actinobacteria" "Actinobacteria"       "Actinomycetales"    "Geodermatophilaceae"
denovo3343 "Bacteria" "Chloroflexi"    "TK10"                 "AKYG885"            "5B-12"              
denovo675  "Bacteria" "Planctomycetes" "vadinHA49"            "DH61"               NA                   
denovo445  "Bacteria" "Acidobacteria"  "[Chloracidobacteria]" "RB41"               "Ellin6075"          
denovo441  "Bacteria" "Actinobacteria" "Actinobacteria"       "Actinomycetales"    "Frankiaceae"        
denovo1731 "Bacteria" "Acidobacteria"  "[Chloracidobacteria]" "RB41"               NA                   
denovo381  "Bacteria" "Acidobacteria"  "Acidobacteriia"       "Acidobacteriales"   "Koribacteraceae"    
```
```
> sample_data(phylo_trim0)[1:10, 1:5]
Sample Data:        [10 samples by 5 sample variables]:
     X.SampleID BarcodeSequence LinkerPrimerSequence         Cave CaveType
S4           S4        AAAAACCG  GAGTTTGATCNTGGCTCAG       Silver Lavatube
V13         V13                                         Valentine Lavatube
C1           C1        AAAAACTC  GAGTTTGATCNTGGCTCAG    Catacombs Lavatube
S10         S10        AAAAACTA  GAGTTTGATCNTGGCTCAG Surface Soil     Soil
GE2         GE2        AAAAAAGG  GAGTTTGATCNTGGCTCAG       Gemini Lavatube
GD3         GD3        AAAAACAG  GAGTTTGATCNTGGCTCAG  Golden Dome Lavatube
L1           L1        AAAAAAGC  GAGTTTGATCNTGGCTCAG Surface Soil     Soil
HC15       HC15        AAAAAACC  GAGTTTGATCNTGGCTCAG Surface Soil     Soil
GD2         GD2        AAAAACAT  GAGTTTGATCNTGGCTCAG  Golden Dome Lavatube
GE13       GE13        AAAAACAA  GAGTTTGATCNTGGCTCAG Surface Soil     Soil
```
```
> # Plots the heat map
> plot_heatmap(phylo_trim0, taxa.order="Phylum", taxa.label="Class", sample.label="Cave", sample.order="CaveType")
Error in order(mdf$Abundance, decreasing = TRUE) : 
  argument 1 is not a vector
```

**Comment by joey711** [comment]
Hi @bioinfonm 

Please update to the latest release version of phyloseq `1.10.0`, and see if this fixes your issue.

```S
source("http://bioconductor.org/biocLite.R")
biocLite("phyloseq")
```

Please post back either way. I will close this issue for now, and re-open if you still have this problem.

Thanks again for your feedback and interest in phyloseq!

joey

**Comment by bioinfonm** [comment]
@joey711 I just updated and I am getting the same error.

ara

**Comment by bioinfonm** [comment]
I am going to try a few other biom and mapping files I have to see if it's a problem with the files.

**Comment by bioinfonm** [comment]
Ok, this is weird. On March 4 I ran the script for creating heatmaps and I got this:
![brown_red_func_comparison_vsd](https://cloud.githubusercontent.com/assets/2340132/6699284/24195418-ccc6-11e4-85c6-0b2250681abd.png)
```
> physeq_func
phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 5673 taxa and 10 samples ]
sample_data() Sample Data:       [ 10 samples by 13 sample variables ]
tax_table()   Taxonomy Table:    [ 5673 taxa by 4 taxonomic ranks ]
```
```
> brown <- subset_samples(physeq_func, Color == "Red")
> red <- subset_samples(physeq_func, Color == "Pink")
> brown_red <- merge_phyloseq(brown,red)
> brown_red
phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 5673 taxa and 5 samples ]
sample_data() Sample Data:       [ 5 samples by 13 sample variables ]
tax_table()   Taxonomy Table:    [ 5673 taxa by 4 taxonomic ranks ]
```
```
> diagdds = phyloseq_to_deseq2(brown_red, ~Color)
converting counts to integer mode
> diagdds
class: DESeqDataSet 
dim: 5673 5 
exptData(0):
assays(1): counts
rownames(5673): 17726 12159 ... 10411 15073
rowData metadata column names(0):
colnames(5): 4609360.3 4609359.3 4609355.3 4609362.3 4609358.3
colData names(13): X.SampleID BarcodeSequence ... Type Description
```
```
> # calculate geometric means prior to estimate size factors
> gm_mean = function(x, na.rm=TRUE){
+ exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
+ }
> geoMeans = apply(counts(diagdds), 1, gm_mean)
> # The effective library size information is called the size factors vector, since the package only
> # needs to know the relative library sizes. If the counts of non-differentially expressed genes in
> #one sample are, on average, twice as high as in another
> diagdds = estimateSizeFactors(diagdds, geoMeans = geoMeans)
> sizeFactors(diagdds)
4609360.3 4609359.3 4609355.3 4609362.3 4609358.3 
 1.515717  1.000000  1.000000  1.000000  1.000000 
> dim(diagdds)
[1] 5673    5
```
> diagdds = DESeq(diagdds, fitType="mean")
using pre-existing size factors
estimating dispersions
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
fitting model and testing
> dim(diagdds)
[1] 5673    5
```
```
> res = results(diagdds)
> res = res[order(res$padj, na.last=NA), ]
> alpha = 0.05
> sigtab = res[(res$padj < alpha), ]
> sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(brown_red)[rownames(sigtab), ], "matrix"))
```
```
> head(sigtab)
      baseMean log2FoldChange    lfcSE      stat      pvalue       padj                   ontology1
10349     41.4      -7.679438 2.713885 -2.829684 0.004659401 0.03369105               Carbohydrates
11749     41.4      -7.679438 2.713885 -2.829684 0.004659401 0.03369105 Clustering-based subsystems
8580      41.4      -7.679438 2.713885 -2.829684 0.004659401 0.03369105 Amino Acids and Derivatives
8657      41.4      -7.679438 2.713885 -2.829684 0.004659401 0.03369105 Amino Acids and Derivatives
9096      41.4      -7.679438 2.713885 -2.829684 0.004659401 0.03369105               Carbohydrates
9597      41.4      -7.679438 2.713885 -2.829684 0.004659401 0.03369105               Carbohydrates
                                        ontology2                                                      ontology3
10349                              Sugar alcohols                                       Ethanolamine_utilization
11749                                           -                                          CBSS-257314.1.peg.752
8580  Lysine, threonine, methionine, and cysteine                                            Lysine_fermentation
8657  Lysine, threonine, methionine, and cysteine                    Threonine_anaerobic_catabolism_gene_cluster
9096              Central carbohydrate metabolism Pyruvate_metabolism_II:_acetyl-CoA,_acetogenesis_from_pyruvate
9597                                 Fermentation                                         Fermentations:_Lactate
                        ontology4
10349 Acetate kinase (EC 2.7.2.1)
11749 Acetate kinase (EC 2.7.2.1)
8580  Acetate kinase (EC 2.7.2.1)
8657  Acetate kinase (EC 2.7.2.1)
9096  Acetate kinase (EC 2.7.2.1)
9597  Acetate kinase (EC 2.7.2.1)
```
```
> dim(sigtab)
[1] 19 10
> brown_red0 = brown_red
> otu_table(brown_red0) <- otu_table(getVarianceStabilizedData(diagdds), taxa_are_rows=TRUE)
```
```
> brown_red0
phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 5673 taxa and 5 samples ]
sample_data() Sample Data:       [ 5 samples by 13 sample variables ]
tax_table()   Taxonomy Table:    [ 5673 taxa by 4 taxonomic ranks ]
```
```
> sample_data(brown_red0)
Sample Data:        [5 samples by 13 sample variables]:
          X.SampleID BarcodeSequence LinkerPrimerSequence BarcodeName    ProjectName Sample        Cave Color   N
4609360.3    4609360              NA                   NA          NA FMD_metagenome   SP43      Spider   Red Med
4609359.3    4609359              NA                   NA          NA FMD_metagenome   SP42      Spider   Red Med
4609355.3    4609355              NA                   NA          NA FMD_metagenome    L48 Lechuguilla   Red Med
4609362.3    4609362              NA                   NA          NA FMD_metagenome   SP46      Spider  Pink Low
4609358.3    4609358              NA                   NA          NA FMD_metagenome   SP41      Spider  Pink Low
             C   Fe Type Description
4609360.3  Low High  FMD     4609360
4609359.3  Low High  FMD     4609359
4609355.3  Low High  FMD     4609355
4609362.3 High  Low  FMD     4609362
4609358.3 High  Low  FMD     4609358
```
```
> # Only looking at the func taxa that are below alpha
> keepOTUs = rownames(res[res$padj < alpha, ])
> brown_red_trim0 = prune_taxa(keepOTUs, brown_red0)
```
```
> plot_heatmap(brown_red_trim0, taxa.order = "ontology1", taxa.label = "ontology2", sample.label = "Color",
+ sample.order = "Color")
Error in order(mdf$Abundance, decreasing = TRUE) : 
  argument 1 is not a vector
In addition: Warning message:
In psmelt(physeq) : The sample variables: 
Sample
 have been renamed to: 
sample_Sample
to avoid conflicts with special phyloseq plot attribute names.
```

**Comment by bioinfonm** [comment]
@joey711 So I tried a bunch of different files and I keep hitting this same error message. I am going to go see if a colleague of mine who is a version back can make the heatplots.

**Comment by bioinfonm** [comment]
Super weird. So I closed R and restarted. I open just one file (MGRAST biom file) and I can get a heatmap no problem.

When I run the biom file from QIIME 1.8 and get the:

```
Error in order(mdf$Abundance, decreasing = TRUE) : 
  argument 1 is not a vector
```

When I switch back to try the MGRAST biom file I get the same error. And every other phyloseq script I open gives me the same error. It's like it's getting stuck.

Once I close out all the scripts and restart R I can make heatmaps expect with that one biom file.

ara

**Comment by joey711** [comment]
What's your OS, version of R, version of BioC, and version of phyloseq now? Did you update yet?

**Comment by bioinfonm** [comment]
@joey711 
OSX 10.10.2

library("phyloseq"); packageVersion("phyloseq")
[1] ‘1.10.0’

$version.string
[1] "R version 3.1.2 (2014-10-31)"

**Comment by joey711** [comment]
Well, since the error is not happening at import, I'm not really concerned about the file format, more about what the R object looks like. That error seems to imply that you have something weird going on. What are the details of the object just prior to calling `plot_heatmap`? Have your tried parameters in `plot_heatmap` for manual ordering of indices? 

Try building a reproducible example of this problem that starts from a fresh R session and your phyloseq object stored as `.RData`, and loaded with the `load` command. This will take out some of these extra details that shouldn't matter and help clarify the problem. From what I can tell, it has nothing to do with the file, but some strange error or inconsistency in your data object that is propagating an error that goes unnoticed until you try to plot it...

**Comment by bioinfonm** [comment]
@joey711 Thanks! I'll give it a try.

**Comment by joey711** [comment]
This is now fixed in `1.13.3+`.
If you don't want to install the devel version (simply wait for the next BioC release), then @michberr 's `detach(reshape)` workaround should also be fine. This is due to a name collision with the old, slow "reshape" package. phyloseq uses the faster "reshape2" package of the same author. They both contain a function called `melt`. This issue is now fixed, as phyloseq explicitly calls `reshape2::melt` in all instances.

Can re-open if this issue resurfaces.

joey
